pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -893,10 +893,48 @@ async def _handle_streamed_model_response(
                 else:
                     messages.append(self._unknown_tool(call.tool_name))
 
-            with _logfire.span('running {tools=}', tools=[t.get_name() for t in tasks]):
-                task_results: Sequence[_messages.Message] = await asyncio.gather(*tasks)
-                messages.extend(task_results)
-            return None, messages
+            tool_messages, stop_run = await self._run_tools(tasks)
+            messages.extend(tool_messages)
+            return _MarkFinalResult(model_response) if stop_run else None, messages
+
+    async def _run_tools(
+        self,
+        tasks: list[asyncio.Task[_messages.Message]],
+    ) -> tuple[list[_messages.Message], _messages.ToolReturn | None]:
+        with _logfire.span('running {tools=}', tools=[t.get_name() for t in tasks]) as span:
+            try:
+                task_results: Sequence[_messages.Message | BaseException] = await asyncio.gather(
+                    *tasks,
+                    return_exceptions=True,  # TODO: Set return_exceptions based on 'eager' vs 'correct'
+                )
+                failures = [
+                    r
+                    for r in task_results
+                    if isinstance(r, BaseException) and not isinstance(r, _exceptions.StopAgentRun)
+                ]
+                if failures:
+                    raise failures[0]  # TODO: raise an exceptiongroup if there is more than 1?
+
+                messages = [r for r in task_results if not isinstance(r, BaseException)]
+                stop_runs = [r for r in task_results if isinstance(r, _exceptions.StopAgentRun)]
+            except _exceptions.StopAgentRun as e:
+                messages = []
+                stop_runs = [e]
+
+            first_tool_return: _messages.ToolReturn | None = None
+            for stop_run in stop_runs:
+                tool_return = _messages.ToolReturn(
```

Comment:
Would it perhaps be better to have some attribute on `ToolReturn` to indicate that it stopped the run, instead of returning `first_tool_return`? So this can be inspected later by the user or elsewhere